### ISSUE TITLE
mg.include not working with document_root other than current directory

### ISSUE THREAD

**Issue by draptik** [civetweb/civetweb#400]
I can't seem to use `mg.include` unless I start civetweb in the current directory using `civetweb -document_root .`.

Here is a minimal example:

`index.lsp`:
```
<!DOCTYPE html>
<html>
<head>
<title>test</title>
</head>
<body>
<p>
<span>Today is:</span>
<? mg.write(os.date("%A")) ?>
</p>
<p>
URI is <?=mg.request_info.uri?>
</p>
<p><? mg.write("Sibling content") ?></p>
<?
mg.include("sibling.lsp")
?>
</body>
</html>
```

`sibling.lsp`:
```
<? mg.write("hello from sibling file") ?>
```

Both files are located in the same folder:
```
ls -alh ~/tmp/www
-rw-r--r-- 1 redheads redheads  257 Jan 16 13:49 index.lsp
-rw-r--r-- 1 redheads redheads   42 Jan 16 13:44 sibling.lsp
```

When I start civetweb in in folder `~/tmp/www` using:
```
cd ~/tmp/www
sudo civetweb -document_root .
```
the `mg.include` works and renders the correct content on the web page.

But if I start civetweb from a different location (ie `~/tmp`) using:
```
cd ~/tmp
sudo civetweb -document_root dev
```
the content of the included file is missing (I have also tried using an absolute path: same error).

Am I missing something obvious?

```
civetweb -I

CivetWeb V1.8
CivetWeb V1.8 - Linux #1 SMP Debian 3.16.39-1 (2016-12-30) (3.16.0-4-amd64) - x86_64
Features: Files HTTPS CGI WebSockets Lua
Lua Version: 502 (Lua 5.2.4)
Version: 1.8
Build: Dec 17 2016
gcc: 4.9.2
Data model: i:2/4/8/8, f:4/8/16, c:1/4, p:8, s:8, t:8
```

**Comment by bel2125** [comment]
My directory structure usually looks like this:
1. I usually have one directory structure "document_root" (called "www" or "web" or "webroot") with all files that can be served *directly*.
2. Then there is a directory for files *not* directly available (called "off" / "offroot" / .. something indicating it is not directly accessible). Typical files would be databases, the Lua preload file, common Lua scripts loaded using "require" or "dofile", content included by scripts in the document_root
3. Then there is a directory "logs" for error.log (and access.log)
4. A working directory ("work")
5. Maybe a separate directory for ssl certificates ("cert"), could somewhere in "off"
6. Maybe a directory for the configuration ("cfg"), could be inside "off" as well
7. On Windows a directory for the binaries ("bin") for civetweb.exe and additional .dll files
8. There could be a different directory for websocket_root as well

In principle it would be possible to have everything in document_root, and use "hide_files_patterns" to hide everything but the files that can be served (part 1 above). It is very likely that you or someone who maintains a running service for several years will expose some files that are not meant to be accessible this way. If you use a dedicated "webroot" folder from the beginning, you do not run into this problem - everything you see in this folder can be seen by http(s).

When using this structure, "index.lua" should be in the document_root, while "sibling.lua" is something you probably do not want to expose directly, so it would be a candidate for "off" (part 2 above). I usually put the path of "off" in a Lua variable (let's call it off_path) in the Lua preload file (config entry lua_preload_file), and include `off_path .. sibling.lua` (Lua concat). If you are sure the file should be accessible directly as well and, thus, located in the document root, you could mg.document_root instead of off_path.

The consideration above is mainly for "Lua scripts" (extension .lua), but you are using "Lua pages" (extension .lp). From the basic idea, Lua pages are closer to "server side includes" (SSI, extension .shtml, https://en.wikipedia.org/wiki/Server_Side_Includes). SSI files are either relative to the parent (with "file=") or relative to the document root (with "virtual=") or absolute or relative to the working directory (with "abspath="). SSI uses different key words / property names for different interpretation of the file path.


> (I have also tried using an absolute path: same error)

How did you do that PRECISELY?
Did you use an absolute path as an argument for document_root when starting the server?
Did you use an absolute path in index.lsp in `mg.include("/home/me/tmp/www/sibling.lsp")`?

**Comment by bel2125** [comment]
Would it be helpful to have a complete example of a web service using the standalone executable?

**Comment by bel2125** [comment]
Don't  you have a header like
```
HTTP/1.0 200 OK
Content-Type: text/html

```
at the beginning of index.lsp (see [page.lp](https://github.com/civetweb/civetweb/blob/master/test/page.lp#L1)), while these lines are missing in sibling.lsp?

**Comment by draptik** [comment]
@bel2125 Thank you for your detailed answer.

>> (I have also tried using an absolute path: same error)

> How did you do that PRECISELY?
> Did you use an absolute path as an argument for document_root when starting the server?
> Did you use an absolute path in index.lsp in mg.include("/home/me/tmp/www/sibling.lsp")?

Thanks for the pointer: Using an absolute path as argument for `document_root` when starting the server does not work. But using an absolute path in `index.lsp` works.

Adding a the header (`HTTP/1.0...`) does not make a difference either way.

> Would it be helpful to have a complete example of a web service using the standalone executable?
> A folder somewhere in the source tree, structured like explained above, with some documentation where to put what files?

Yes, this would be helpful.

So, my question remains: 
- Is this the expected behavior (`mg.include` only works when `document_root` is the current folder `.`)
- OR am I doing something wrong
- OR is this a bug?

**Comment by bel2125** [comment]
> Yes, this would be helpful.

I will put a full application somewhere in the repository, with all parts MIT licensed. But this might take some days or weeks. I can't just take something that may contain sensitive data or know how, incompatible licenses, anything covered by a NDA, or anything extremely specific or something with too much need of explanation.

> So, my question remains:
>    Is this the expected behavior (mg.include only works when document_root is the current folder .)
>    OR am I doing something wrong
>    OR is this a bug?

No, no and no - it's an enhancement, but let's go through this in detail:

> (mg.include only works when document_root is the current folder .)

It works if you use a proper path in mg.include, right?
> But using an absolute path in index.lsp works.

I added this test yesterday ([page3.lp](https://github.com/civetweb/civetweb/blob/master/test/page3.lp)) - it includes a Lua page ([page2.lp](https://github.com/civetweb/civetweb/blob/master/test/page2.lp)) in the same directory independent from the document_root setting and independent from a sub-folder within document_root. 
The `string.gsub(mg.script_name, ...)` expression works, but I think it is somewhat "ugly". I will try to find a more "beautiful" one - maybe we need an additional variable in the Lua environment.

>    OR is this a bug?

I know, that's still not a direct answer to *"is it a bug?"*.
I don't think it's a bug in the implementation, but a lack in the specification. The [User Manual](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md) currently just states:

`mg.include(path)           -- sources another Lua file`

But does not specify how this path must be interpreted. In contrast, the same user manual does specify this for [SSI includes](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#ssi_pattern-shtmlshtm). It defines
```
virtual ... Path is relative to web server root
abspath ... Path is absolute or relative to web server working dir
file ... Path is relative to current document
```

>    OR am I doing something wrong

Since the specification is not precise enough here, I would not say "you are wrong".

The behavior of `mg.include` is equivalent to `abspath`. Changing this would break compatibility, so I would not do this. However, I must append this information to the user manual.

However, I do see the need for Lua includes equivalent to "virual" and "file" in SSI as well.
This is why I do consider this an enhancement.

**Comment by bel2125** [comment]
How would/could/should this feature look like:

One option would be to add a new optional parameter `pathtype` to `mg.include(path, pathttype)`, to choose if it is a `virtual`, `file` or `abspath` kind of path, with the default `abspath` to retain compatibility.
Would this be suitable for you?

This can be done in C (by me in V1.10), or in a Lua in a pre-loaded library (by you now). The Lua library would basically do some string gymnastics (like the gsub above) to get the absolute path from the relative path. 

One thing I still makes me wonder:

> Adding a the header (HTTP/1.0...) does not make a difference either way.

How did you test this? 
Could you try page2.lp and page3.lp?

**Comment by draptik** [comment]
> One option would be to add a new optional parameter pathtype to mg.include(path, pathttype), to choose if it is a virtual, file or abspath kind of path, with the default abspath to retain compatibility.
> Would this be suitable for you?

Sounds like a good idea to me.

> This can be done in C (by me in V1.10)

Cool!

**Comment by bel2125** [comment]
The documentation is updated now.

    mg.get_mime_type(filename) -- get MIME type of a file
    mg.send_file(filename)     -- send a file, including MIME type

Have the same lack of specification. Probably another solution would be not to add an optional parameter everywhere, but to have a helper function `mg.rel_to_abs_path(rel_path, path_type)`. Or the script path in a string variable, so you can concat yourself.

**Comment by draptik** [comment]
I'm guessing: Most consumers of `mg.include` will be happy if `virtual` is the default in the future.

**Comment by bel2125** [comment]
> I'm guessing: Most consumers of mg.include will be happy if virtual is the default in the future.

That might make current customers quite unhappy, since I would brake their existing applications.

**Comment by draptik** [comment]
Yes, sorry. That was just plain wrong. I normally work with enterprise stuff. Just forget the answer about 'virtual'

**Comment by bel2125** [comment]
Still, one open topic is the HTTP header - the lines
```
HTTP/1.0 200 OK
Content-Type: text/html

```
on top of every *.lp file. You don't have that?
What client (browser?) did you use?

**Comment by draptik** [comment]
personally? Chrom*, FF (VM: IE)

**Comment by draptik** [comment]
No, I don't have the header. All test works (civetweb / archlinux / ubuntu / raspbian)

**Comment by bel2125** [comment]
> personally? Chrom*, FF (VM: IE)

I meant for the test of the *.lp pages.
It's the responsibility of the *.lp files to sent the HTTP response headers before the actual content. If you don't do that, it's a violation of HTTP. The client might ignore this violation, or reject the entire response.

**Comment by bel2125** [comment]
> No, I don't have the header. All test works (civetweb / archlinux / ubuntu / raspbian)

I usually use the "Fiddler2" web debugging tool to check for protocol violations. 
I use it on Windows - I don't know if there is a Linux version as well, but since it runs on the same machine as the client (web browser) I don't need it for Linux even if the server runs on Linux.
Probably there are other tools and browser plugins that do the same as well.

**Comment by bel2125** [comment]
This content should work for index.lsp (tested on Windows, not yet on Linux):
```Lua
HTTP/1.1 200 OK
Content-Type: text/html
Connection: close

<!DOCTYPE html>
<html>
<head>
<title>test</title>
</head>
<body>
<p>
<span>Today is:</span>
<? mg.write(os.date("%A")) ?>
</p>
<p>
URI is <?=mg.request_info.uri?>
</p>
<p><? mg.write("Sibling content") ?></p>
<?
mg.include(mg.script_name:gsub("%/[^/]*$", "/") .. "sibling.lsp")
?>
</body>
</html>

```
You DO need the `HTTP/1.0 200 OK` line followed by an empty line in index.lsp, and you MUST NOT have this line in sibling.lsp - it would be a HTTP standard violation otherwise. Browsers try to work with this violation, but FireFox displayed the content differently with and without header.
Fiddler shows an error if the header is missing.

**Comment by draptik** [comment]
@bel2125 Thank you very much: Your solution works!

To summarize:

1. Adding header to the calling LSP file (`index.lsp`):
```
HTTP/1.1 200 OK
Content-Type: text/html
Connection: close
```
2. Replace
```
mg.include("sibling.lsp")
```
with
```
mg.include(mg.script_name:gsub("%/[^/]*$", "/") .. "sibling.lsp")
```

Great!

**Comment by bel2125** [comment]
In addition to the "pure Lua" solution using string substitution, the `mg.include` function now supports a second argument to specify if the path type ("abs", "rel"/"file" or "virt[ual]", see [User Manual](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#lua-scripts-and-lua-server-pages) and `test/page3?.lp`). 

This is already available in the 1.10 release from September 2017.
So I think, we can close this issue now.
